core/trino-main/src/main/java/io/trino/sql/planner/optimizations/DeterminePartitionCount.java

```diff
@@ -0,0 +1,247 @@
+    {
+        int maxHashPartitionCount = getMaxHashPartitionCount(session);
+        long minInputRowsPerTask = getMinInputRowsPerTask(session);
+
+        if (minInputRowsPerTask == 0) {
+            return Optional.empty();
+        }
+
+        // Skip for expanding plan nodes like CROSS JOIN or UNNEST which can substantially increase the amount of data.
+        if (isInputMultiplyingPlanNodesPresent(plan)) {
+            return Optional.empty();
+        }
+
+        CachingStatsProvider statsProvider = new CachingStatsProvider(statsCalculator, session, types, tableStatsProvider);
+
+        double sourceTablesRowCount = getSourcePlanNodesRowCount(plan, statsProvider);
+        double expandingNodesMaxRowCount = getExpandingNodesMaxRowCount(plan, statsProvider);
+        if (isNaN(sourceTablesRowCount) || isNaN(expandingNodesMaxRowCount)) {
+            return Optional.empty();
+        }
+
+        int partitionCount = max(
+                getPartitionsFromRowCount(sourceTablesRowCount, minInputRowsPerTask),
+                getPartitionsFromRowCount(expandingNodesMaxRowCount, minInputRowsPerTask));
+
+        if (partitionCount >= maxHashPartitionCount) {
+            return Optional.empty();
+        }
+
+        log.debug("Estimated remote exchange partition count for query %s is %s", session.getQueryId(), partitionCount);
+        return Optional.of(partitionCount);
+    }
+
+    private static int getPartitionsFromRowCount(double rowCount, long minInputRowsPerTask)
+    {
+        return max((int) (rowCount / minInputRowsPerTask), 1);
+    }
+
+    private static boolean isInputMultiplyingPlanNodesPresent(PlanNode root)
+    {
+        return PlanNodeSearcher.searchFrom(root)
+                .where(DeterminePartitionCount::isInputMultiplyingPlanNode)
+                .matches();
+    }
+
+    private static boolean isInputMultiplyingPlanNode(PlanNode node)
+    {
+        if (node instanceof UnnestNode) {
+            return true;
+        }
+
+        if (node instanceof JoinNode crossJoinNode && ((JoinNode) node).isCrossJoin()) {
+            // If any of the input node is scalar then there's no need to skip cross join
+            return !isAtMostScalar(crossJoinNode.getRight()) && !isAtMostScalar(crossJoinNode.getLeft());
+        }
+
+        return false;
+    }
+
+    private static double getExpandingNodesMaxRowCount(PlanNode root, StatsProvider statsProvider)
+    {
+        List<PlanNode> expandingNodes = PlanNodeSearcher.searchFrom(root)
+                .where(DeterminePartitionCount::isExpandingPlanNode)
+                .findAll();
+
+        return expandingNodes.stream()
+                .mapToDouble(node -> statsProvider.getStats(node).getOutputRowCount())
+                .max()
+                .orElse(0);
+    }
+
+    private static boolean isExpandingPlanNode(PlanNode node)
+    {
+        return node instanceof JoinNode
+                || node instanceof UnionNode
```

Comment:
Is Union expanding?

Reply:
It is based on the assumption that you can have union over multiple joins which can increase the rows exchange remotely

Reply:
The UNION operation is not expanding per say. It simply coalesces inputs from multiple sources, but it doesn't introduce any new data. It's a little unclear why is it require a special treatment.